File: answers/leetcode/Populating Next Right Pointers in Each Node.cpp
//@result 14 / 14 test cases passed. Status: Accepted Runtime: 24 ms Submitted: 0 minutes ago You are here! Your runtime beats 42.22% of cpp submissions.

class Solution {
public:
    void connect(TreeLinkNode *root) {
		TreeLinkNode *level = root;
		while (level) {
			TreeLinkNode *current_parent = level;
			TreeLinkNode *current = NULL;
			while (current_parent) {
				if (current_parent->left) {
					current = current_parent->left;
					current->next = current_parent->right;
				}
				if (current_parent->right) {
					current = current_parent->right;
				}
				// if no children
				if (current && current_parent->next) {
					current->next = current_parent->next->left ? current_parent->next->left : current_parent->next->right;
				}
				current_parent = current_parent->next;
			}
			level = level->left ? level->left : level->right;
		}
    }
};

File: answers/other/coding/Populating Next Right Pointers in Each Node.py
# Definition for binary tree with next pointer.
# class TreeLinkNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution(object):
	def connect(self, root):
		"""
		:type root: TreeLinkNode
		:rtype: nothing
		"""
		
		if root:
			outgoing = None
			if root.left:
				if root.right:
					root.left.next = root.right
				else:
					outgoing = root.left
			if root.right:
				outgoing = root.right
			if outgoing:
				current = root.next
				while current:
					if current.left:
						outgoing.next = current.left
						break
					elif current.right:
						outgoing.next = current.right
						break
					current = current.next
			self.connect(root.left)
			self.connect(root.right)
